allennlp/data/dataset_readers/dataset_utils/text2sql_utils.py

```diff
@@ -78,6 +82,25 @@ def clean_and_split_sql(sql: str) -> List[str]:
             sql_tokens.extend(split_table_and_column_names(token))
     return sql_tokens
 
+def resolve_primary_keys_in_schema(sql_tokens: List[str],
+                                   schema: Dict[str, List[TableColumn]]) -> List[str]:
+    """
+    Some examples in the text2sql datasets use ID as a column reference to the
+    column of a table which has a primary key. This causes problems if you are trying
+    to constrain a grammar to only produce the column names directly, because you don't
+    know what ID refers to. So instead of dealing with that, we just replace it.
+    """
+    primary_keys_for_tables = {name: max(columns, key=lambda x: x.is_primary_key).name
```

Comment:
Don't you want some `if` in here?  What happens if no column is a primary key?

Reply:
If a column doesn't have a primary key, there will be no ID reference to it, so it's fine because the max will just take an arbitrary one.